frontend/src/utilities/documents.ts

```diff
@@ -84,16 +93,49 @@ export default readonly(state);
 registerResponseHandler(ResponseType.UpdateOpenDocumentsList, (responseData: Response) => {
 	const documentListData = responseData as UpdateOpenDocumentsList;
 	if (documentListData) {
-		state.documents = documentListData.open_documents;
-		state.title = state.documents[state.activeDocumentIndex];
+		state.documents = documentListData.open_documents.map((name) => {
+			// Can't guarantee an exact match since it is going off of names
+			// This is ok though since it is only used to mark a document as unsaved
+			const probableMatch = state.documents.find((doc) => doc.name === name);
+			if (probableMatch) {
+				return { name, unsaved: probableMatch.unsaved, layerTreeHeight: probableMatch.layerTreeHeight };
+			}
+			return { name, unsaved: false, layerTreeHeight: 0 };
+		});
+		state.title = state.documents[state.activeDocumentIndex].name;
 	}
 });
 
 registerResponseHandler(ResponseType.SetActiveDocument, (responseData: Response) => {
 	const documentData = responseData as SetActiveDocument;
 	if (documentData) {
 		state.activeDocumentIndex = documentData.document_index;
-		state.title = state.documents[state.activeDocumentIndex];
+		state.title = state.documents[state.activeDocumentIndex].name;
+	}
+});
+
+// Used to figure out if the document has been modified since the last save
+// Pitfalls: This will not detect if a new effect has been applied but just if the hierarchy has been modified
```

Comment:
This is a nice attempt working with what you have in the browser, since it has very little "context" about the state of the application. Although I believe the only way to do this "correctly" would be to edit the Rust code which has the entire application state and contextual information. Let's pair program this so I can show you the Rust codebase and we can make this work based on any change that has occurred since the document was saved.